modules/launcher/src/main/java/org/mule/runtime/module/launcher/ApplicationDescriptorFactory.java

```diff
@@ -97,6 +104,30 @@ public ApplicationDescriptor create(File artifactFolder) throws ArtifactDescript
         return desc;
     }
 
+    private List<ApplicationPluginDescriptor> getAllApplicationPlugins(Set<ApplicationPluginDescriptor> plugins)
+    {
+        final List<ApplicationPluginDescriptor> result = new LinkedList<>(applicationPluginRepository.getContainerApplicationPluginDescriptors());
+        result.addAll(plugins);
+        return result;
+    }
+
+    private void verifyPluginExportedPackages(List<ApplicationPluginDescriptor> plugins)
+    {
+        final HashMap<String, String> exportedPackages = new HashMap<>();
+
+        for (ApplicationPluginDescriptor plugin : plugins)
+        {
+            for (String exportedPackage : plugin.getClassLoaderFilter().getExportedClassPackages())
+            {
+                if (exportedPackages.containsKey(exportedPackage))
+                {
+                    throw new DuplicateExportedPackageException(String.format("Package '%s' is exported on plugins '%s' and '%s'", exportedPackage, exportedPackages.get(exportedPackage), plugin.getName()));
```

Comment:
Maybe a different check/wording can be done to account for the case that a same package is exported by more than 2 plugins, otherwise another deployment would be needed to detect that case.